Sift algorithm

ShubhamGupta577 opened the pull request:
# Description

- In this script, we implemented SIFT(Scale Invariant Feature Transform) algorithm of Open CV using Python.
- This algorithm is used to detect the various feature of an image.

Fixes #967 


## Type of change
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works

ShubhamGupta577 commented:
@HarshCasper, Please review this PR.

HarshCasper commented on SIFT Algorithm/SIFT_Algorithm.py:
```diff
@@ -0,0 +1,36 @@
+print("Number of Keypoints Detected In The Image: ", len(kp))
+
+#Display the final output image
+cv2_imshow('sift_keypoints.jpg',img)
```
Can we have a dynamic output image name as well?

HarshCasper commented on SIFT Algorithm/SIFT_Algorithm.py:
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+
+# Load the image
+path=input("Enter the path of the image: ")
+image = cv2.imread(path)
+
+# Convert the image to RGB
+image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+
+# Convert the image to gray scale
+gray = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
+
+#Display the input image
+cv2.imshow('Input image.jpg',image)
+
+#Detect keypoints and Create Descriptor
+sift = cv2.xfeatures2d.SIFT_create()
+
+kp, dp = sift.detectAndCompute(gray, None)
+
+img=cv2.drawKeypoints(gray,kp,flags = cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+
+# Print the number of keypoints detected in the image
+print("Number of Keypoints Detected In The Image: ", len(kp))
```
Put all of the code in a modular function and call the function from a driver code

sohamsshah commented:
@ShubhamGupta577 IMO pure and direct algorithm scripts are less preferable. If you can create a utilitarian script by making use of this algorithm, I suppose it can provide more value to the project. You can search for applications of this algorithm and then implement. Idea is to have scripts that are helpful in some way to the developers/users. 

I hope you understand, thanks.

ShubhamGupta577 commented:
@sohamsshah, I understand your concern but IMO the script which I added is more clear and helpful for the users/developers because whoever needs this script is directly available here otherwise they have to extract the useful code from this script which will be more tedious.

ShubhamGupta577 commented on SIFT Algorithm/Readme.md:
```diff
@@ -0,0 +1,16 @@
+
+## Output
+
+<img src="https://i.ibb.co/SJ1rfC3/Input-Image.png" width=450/> <img src="https://i.ibb.co/VHgMqWN/Output-image.png" width=450/>
```
I had to resize the image for the readme and by using markdown I was not able to achieve it. So I had used HTML tags for the same purpose.

ShubhamGupta577 commented:
@HarshCasper, Requested changes had done. Please review the script.

sohamsshah commented:
> @sohamsshah, I understand your concern but IMO the script which I added is more clear and helpful for the users/developers because whoever needs this script is directly available here otherwise they have to extract the useful code from this script which will be more tedious.

I appreciate your efforts for working on this. But, this script is pure algorithm implementation. SIFT is a really good algorithm which is used at various CV applications, but pure algorithm implementations aren't applicable to the niche of Rotten Scripts.

As I said, if you can bring some use-case script that uses one or more CV concepts to solve a problem, it is readily welcomed. I hope you understand.

ShubhamGupta577 commented:
@sohamsshah, I understand what you say, As SIFT is invariance to image scale and rotation for the feature detection of an image. So as an addition to the script can we add the code to
- Generate a test image by adding Scale invariance and rotational invariance to the input image which changes the alignment and resolution of the image.
- Then apply  **Brute Force Matching Algorithm of Open CV** for matching the features of input and test image. This will show the application of SIFT.

If it makes our script useful, please let me know.

sohamsshah commented:
> @sohamsshah, I understand what you said, As SIFT is invariance to image scale and rotation for the feature detection of an image. So as an addition to the script can we add the code to
> 
> * Generate a test image by adding Scale invariance and rotational invariance to the input image which changes the alignment and resolution of the image.
> * Then apply  **Brute Force Matching Algorithm of Open CV** for matching the features of input and test image. This will show the application of SIFT.
> 
> If it makes our script useful, please let me know.

Okay, sounds good. You can go ahead and rename the script accordingly. Also, try to produce enough examples and descriptive README for the same for the users to understand the script's utility clearly.

Thanks for working on this

ShubhamGupta577 commented:
@sohamsshah, I had added the code of Brute Force Matcher for matching the features of the input and test image and also updated the readme file.
Please review this PR.

sohamsshah reviewed (changes requested):
Good work @ShubhamGupta577 !

Few small suggestions from my side:
- I see that the `SIFT ALGORITHM` file is put in the root folder of the repo. Can you shift this directory under Python? 
- I would suggest to add 2-3 more `input` `output` examples to the Readme.
- You can also share the CV pipeline (a brief flow of how various CV operations lead to the final output) in the README.md

After that, its good to merge. 
Again thanks for working on this :)

sohamsshah reviewed (approved):
Great work @ShubhamGupta577 
LGTM @HarshCasper

vybhav72954 commented on Python/SIFT Algorithm/SIFT_Algorithm.py:
```diff
@@ -0,0 +1,74 @@
+import cv2
+import numpy as np
+
+def siftAlgorithm():
```
Missing a docstring here

vybhav72954 commented on Python/SIFT Algorithm/SIFT_Algorithm.py:
```diff
@@ -0,0 +1,74 @@
+    # Print total number of matching points between the training and query images
+    print("\nNumber of Matching Keypoints Between The Training and Query Images: ", len(matches))
+
+#Driver Code
```
Can we trigger this using the proper PIP guide?

Something like
```
if __name__ == "__main__"
```
